File: backend/src/services/tool-executor.ts
import { randomUUID } from 'node:crypto';

import { FieldPath, Timestamp } from 'firebase-admin/firestore';
import { calendar_v3, google } from 'googleapis';

import { env } from '../config/env.js';
import { logger } from '../lib/logger.js';
import { adminFirestore } from './firebase-admin.js';

type ToolResult = Record<string, unknown>;

type ReminderStatus = 'pending' | 'fired' | 'dismissed' | 'snoozed';
type MemoryCategory = 'preferences' | 'facts' | 'habits' | 'health' | 'routines';

export class ToolExecutor {
  constructor(private readonly userId: string) {}

  async execute(toolName: string, input: Record<string, unknown>): Promise<ToolResult> {
    switch (toolName) {
      case 'set_reminder':
        return this.setReminder(input);
      case 'list_reminders':
        return this.listReminders(input);
      case 'cancel_reminder':
        return this.cancelReminder(input);
      case 'create_calendar_event':
        return this.createCalendarEvent(input);
      case 'get_upcoming_events':
        return this.getUpcomingEvents(input);
      case 'store_memory':
        return this.storeMemory(input);
      case 'query_memory':
        return this.queryMemory(input);
      case 'analyze_nutrition':
        return this.analyzeNutrition(input);
      case 'get_user_context':
        return this.getUserContext(input);
      default:
        throw new Error(`Unsupported tool: ${toolName}`);
    }
  }

  private userRef() {
    return adminFirestore().collection('users').doc(this.userId);
  }

  private remindersRef() {
    return this.userRef().collection('reminders');
  }

  private memoriesRef() {
    return this.userRef().collection('memories');
  }

  private nutritionLogsRef() {
    return this.userRef().collection('nutrition_logs');
  }

  private async setReminder(input: Record<string, unknown>): Promise<ToolResult> {
    const message = String(input.message ?? '').trim();
    const delayMinutes = Number(input.delay_minutes ?? 0);
    const priority = String(input.priority ?? 'normal');

    if (!message || !Number.isFinite(delayMinutes) || delayMinutes <= 0) {
      throw new Error('set_reminder requires a message and a positive delay_minutes.');
    }

    const id = randomUUID();
    const triggerAt = new Date(Date.now() + delayMinutes * 60_000);
    await this.remindersRef().doc(id).set({
      message,
      trigger_at: triggerAt.toISOString(),
      status: 'pending',
      priority,
      created_via: 'voice',
      snooze_count: 0,
      created_at: new Date().toISOString(),
    });

    return {
      reminder_id: id,
      message,
      trigger_at: triggerAt.toISOString(),
      status: 'pending',
      priority,
    };
  }

  private async listReminders(input: Record<string, unknown>): Promise<ToolResult> {
    const statusFilter = String(input.status_filter ?? 'pending') as
      | 'pending'
      | 'fired'
      | 'all';

    let query:
      | FirebaseFirestore.Query<FirebaseFirestore.DocumentData>
      | FirebaseFirestore.CollectionReference<FirebaseFirestore.DocumentData> =
      this.remindersRef().orderBy('trigger_at');

    if (statusFilter !== 'all') {
      query = query.where('status', '==', statusFilter);
    }

    const snapshot = await query.get();
    return {
      reminders: snapshot.docs.map((doc) => ({
        reminder_id: doc.id,
        ...doc.data(),
      })),
    };
  }

  private async cancelReminder(input: Record<string, unknown>): Promise<ToolResult> {
    const reminderId = String(input.reminder_id ?? '').trim();
    if (!reminderId) {
      throw new Error('cancel_reminder requires reminder_id.');
    }

    await this.remindersRef().doc(reminderId).update({
      status: 'dismissed',
      dismissed_at: new Date().toISOString(),
    });

    return {
      reminder_id: reminderId,
      status: 'dismissed',
    };
  }

  private async storeMemory(input: Record<string, unknown>): Promise<ToolResult> {
    const key = String(input.key ?? '').trim();
    const value = String(input.value ?? '').trim();
    const category = String(input.category ?? 'facts') as MemoryCategory;

    if (!key || !value) {
      throw new Error('store_memory requires key and value.');
    }

    const existing = await this.memoriesRef().where('key', '==', key).limit(1).get();
    const id = existing.docs[0]?.id ?? randomUUID();
    const now = new Date().toISOString();

    await this.memoriesRef().doc(id).set(
      {
        key,
        value,
        category,
        source: 'voice',
        created_at: existing.docs[0]?.get('created_at') ?? now,
        updated_at: now,
      },
      { merge: true },
    );

    return {
      memory_id: id,
      key,
      value,
      category,
    };
  }

  private async queryMemory(input: Record<string, unknown>): Promise<ToolResult> {
    const queryText = String(input.query ?? '').trim().toLowerCase();
    const categoryFilter = String(input.category_filter ?? 'all');
    if (!queryText) {
      throw new Error('query_memory requires query.');
    }

    let query:
      | FirebaseFirestore.Query<FirebaseFirestore.DocumentData>
      | FirebaseFirestore.CollectionReference<FirebaseFirestore.DocumentData> =
      this.memoriesRef().orderBy(FieldPath.documentId());

    if (categoryFilter !== 'all') {
      query = query.where('category', '==', categoryFilter);
    }

    const snapshot = await query.get();
    const allMemories = snapshot.docs
      .map((doc) => ({
        memory_id: doc.id,
        ...doc.data(),
      }));

    const matches = (queryText
      ? allMemories.filter((memory) => {
        const haystack = `${String(memory.key)} ${String(memory.value)}`.toLowerCase();
        return haystack.includes(queryText);
      })
      : allMemories)
      .slice(0, 10);

    return { matches };
  }

  private async createCalendarEvent(input: Record<string, unknown>): Promise<ToolResult> {
    const title = String(input.title ?? '').trim();
    const startTime = String(input.start_time ?? '').trim();
    const endTime = input.end_time ? String(input.end_time) : undefined;

    if (!title || !startTime) {
      throw new Error('create_calendar_event requires title and start_time.');
    }

    const calendar = await this.getCalendarClient();
    if (!calendar) {
      return {
        configured: false,
        message: 'Google Calendar is not configured for this user.',
      };
    }

    const response = await calendar.events.insert({
      calendarId: 'primary',
      requestBody: {
        summary: title,
        description: input.description ? String(input.description) : undefined,
        location: input.location ? String(input.location) : undefined,
        start: {
          dateTime: startTime,
        },
        end: {
          dateTime: endTime ?? new Date(new Date(startTime).getTime() + 30 * 60_000).toISOString(),
        },
      },
    });

    return {
      configured: true,
      event_id: response.data.id,
      html_link: response.data.htmlLink,
      status: response.data.status,
    };
  }

  private async getUpcomingEvents(input: Record<string, unknown>): Promise<ToolResult> {
    const hoursAhead = Number(input.hours_ahead ?? 24);
    const calendar = await this.getCalendarClient();
    if (!calendar) {
      return {
        configured: false,
        events: [],
      };
    }

    const now = new Date();
    const end = new Date(now.getTime() + hoursAhead * 60 * 60_000);
    const response = await calendar.events.list({
      calendarId: 'primary',
      timeMin: now.toISOString(),
      timeMax: end.toISOString(),
      singleEvents: true,
      orderBy: 'startTime',
      maxResults: 20,
    });

    return {
      configured: true,
      events: (response.data.items ?? []).map((event) => ({
        id: event.id,
        title: event.summary,
        start_time: event.start?.dateTime ?? event.start?.date,
        end_time: event.end?.dateTime ?? event.end?.date,
        location: event.location,
      })),
    };
  }

  private async analyzeNutrition(input: Record<string, unknown>): Promise<ToolResult> {
    const ocrText = String(input.ocr_text ?? '').trim();
    if (!ocrText) {
      throw new Error('analyze_nutrition requires ocr_text.');
    }

    const extractNumber = (pattern: RegExp): number | null => {
      const match = pattern.exec(ocrText);
      if (!match) return null;
      return Number(match[1]);
    };

    const calories = extractNumber(/calories\s+(\d+)/i);
    const protein = extractNumber(/protein\s+(\d+)/i);
    const sugar = extractNumber(/sugars?\s+(\d+)/i);
    const sodium = extractNumber(/sodium\s+(\d+)/i);
    const quantity = typeof input.quantity === 'number' ? input.quantity : 1;

    const concerns = [
      if ((sugar ?? 0) * quantity >= 20) 'high sugar',
      if ((sodium ?? 0) * quantity >= 600) 'high sodium',
      if ((protein ?? 0) * quantity <= 5) 'low protein',
    ];

    const recommendation =
        concerns.contains('high sugar') || concerns.contains('high sodium')
          ? 'moderate'
          : 'eat';

    const logId = randomUUID();
    await this.nutritionLogsRef().doc(logId).set({
      ocr_text: ocrText,
      occasion: input.occasion ?? null,
      quantity,
      is_cheat_meal: input.is_cheat_meal ?? null,
      analysis: `Calories: ${calories ?? 'unknown'}, protein: ${protein ?? 'unknown'}g, sugar: ${sugar ?? 'unknown'}g, sodium: ${sodium ?? 'unknown'}mg.`,
      recommendation,
      timestamp: new Date().toISOString(),
    });

    return {
      nutrition_log_id: logId,
      calories,
      protein_grams: protein,
      sugar_grams: sugar,
      sodium_mg: sodium,
      quantity,
      concerns,
      recommendation,
    };
  }

  private async getUserContext(input: Record<string, unknown>): Promise<ToolResult> {
    const includeMemories = input.include_memories !== false;
    const includeReminders = input.include_reminders !== false;
    const includeEvents = input.include_events !== false;

    const [memories, reminders, events] = await Promise.all([
      includeMemories
        ? this.queryMemory({ query: '', category_filter: 'all' }).catch(() => ({ matches: [] }))
        : Promise.resolve({ matches: [] }),
      includeReminders ? this.listReminders({ status_filter: 'pending' }).catch(() => ({ reminders: [] })) : Promise.resolve({ reminders: [] }),
      includeEvents ? this.getUpcomingEvents({ hours_ahead: 24 }).catch(() => ({ events: [] })) : Promise.resolve({ events: [] }),
    ]);

    return {
      user_id: this.userId,
      memories: memories.matches ?? [],
      reminders: reminders.reminders ?? [],
      upcoming_events: events.events ?? [],
    };
  }

  private async getCalendarClient(): Promise<calendar_v3.Calendar | null> {
    if (!env.GOOGLE_CLIENT_ID || !env.GOOGLE_CLIENT_SECRET || !env.GOOGLE_REDIRECT_URI) {
      return null;
    }

    const tokensDoc = await this.userRef()
      .collection('integrations')
      .doc('google_calendar')
      .get();

    if (!tokensDoc.exists) {
      return null;
    }

    const tokens = tokensDoc.data();
    if (!tokens) {
      return null;
    }

    const auth = new google.auth.OAuth2(
      env.GOOGLE_CLIENT_ID,
      env.GOOGLE_CLIENT_SECRET,
      env.GOOGLE_REDIRECT_URI,
    );

    auth.setCredentials({
      access_token: typeof tokens.access_token === 'string' ? tokens.access_token : undefined,
      refresh_token: typeof tokens.refresh_token === 'string' ? tokens.refresh_token : undefined,
      expiry_date:
        typeof tokens.expiry_date === 'number'
          ? tokens.expiry_date
          : undefined,
    });

    return google.calendar({ version: 'v3', auth });
  }
}

export const fetchDueReminders = async (): Promise<
  Array<{ userId: string; reminderId: string; data: Record<string, unknown> }>
> => {
  const nowIso = new Date().toISOString();
  const snapshot = await adminFirestore()
    .collectionGroup('reminders')
    .where('status', '==', 'pending')
    .where('trigger_at', '<=', nowIso)
    .get();

  return snapshot.docs.map((doc) => {
    const userId = doc.ref.parent.parent?.id;
    if (!userId) {
      throw new Error(`Unable to resolve user for reminder ${doc.id}`);
    }
    return {
      userId,
      reminderId: doc.id,
      data: doc.data(),
    };
  });
};

export const markReminderFired = async (
  userId: string,
  reminderId: string,
): Promise<void> => {
  await adminFirestore()
    .collection('users')
    .doc(userId)
    .collection('reminders')
    .doc(reminderId)
    .update({
      status: 'fired' satisfies ReminderStatus,
      fired_at: Timestamp.now().toDate().toISOString(),
    });
};

export const listUserFcmTokens = async (userId: string): Promise<string[]> => {
  const userDoc = await adminFirestore().collection('users').doc(userId).get();
  const tokens = userDoc.get('fcm_tokens');
  if (!Array.isArray(tokens)) {
    return [];
  }
  return tokens.filter((value): value is string => typeof value === 'string');
};

export const logToolFailure = (toolName: string, error: unknown): void => {
  logger.error('Tool execution failed', {
    toolName,
    error: error instanceof Error ? error.message : String(error),
  });
};

File: backend/src/services/tool_executor.py
"""
ToolExecutor — implements all tools.
"""

from __future__ import annotations

import asyncio
import re
from datetime import datetime, timedelta, timezone
from typing import Any
from uuid import uuid4

from google.cloud import firestore as fs
from google.cloud.firestore_v1.base_query import FieldFilter

from ..config.settings import settings
from ..lib.logger import logger
from .firebase import admin_firestore, admin_messaging
from .google_calendar_connector import GoogleCalendarConnector

ToolResult = dict[str, Any]

TOOL_TIMEOUT_S = settings.VOICE_TOOL_TIMEOUT_S


# runs sync functions with timeout
async def _run(fn, *args, **kwargs):
    return await asyncio.wait_for(asyncio.to_thread(fn, *args, **kwargs), timeout=TOOL_TIMEOUT_S)


class ToolExecutor:
    def __init__(self, user_id: str) -> None:
        self._user_id = user_id

    def _db(self) -> fs.Client:
        return admin_firestore()

    def _user_ref(self) -> fs.DocumentReference:
        return self._db().collection("users").document(self._user_id)

    def _reminders_ref(self) -> fs.CollectionReference:
        return self._user_ref().collection("reminders")

    def _memories_ref(self) -> fs.CollectionReference:
        return self._user_ref().collection("memories")

    def _nutrition_logs_ref(self) -> fs.CollectionReference:
        return self._user_ref().collection("nutrition_logs")

    async def execute(self, tool_name: str, input_data: dict[str, Any]) -> ToolResult:
        dispatch: dict[str, Any] = {
            "set_reminder": self._set_reminder,
            "list_reminders": self._list_reminders,
            "cancel_reminder": self._cancel_reminder,
            "create_calendar_event": self._create_calendar_event,
            "get_upcoming_events": self._get_upcoming_events,
            "store_memory": self._store_memory,
            "query_memory": self._query_memory,
            "analyze_nutrition": self._analyze_nutrition,
            "get_user_context": self._get_user_context,
            "ask_clarification": self._ask_clarification,
        }
        handler = dispatch.get(tool_name)
        if handler is None:
            logger.warn("Tool: unknown tool requested", {"tool": tool_name, "user_id": self._user_id})
            return {"error": f"Unknown tool: {tool_name}"}

        import time as _time
        _start = _time.monotonic()
        logger.debug(f"Tool: executing {tool_name}", {
            "user_id": self._user_id,
            "input_keys": list(input_data.keys()),
        })
        try:
            result = await handler(input_data)
            _ms = int((_time.monotonic() - _start) * 1000)
            logger.info(f"Tool: {tool_name} OK", {
                "user_id": self._user_id,
                "duration_ms": _ms,
                "result_keys": list(result.keys()) if isinstance(result, dict) else "non-dict",
            })
            return result
        except Exception as exc:
            _ms = int((_time.monotonic() - _start) * 1000)
            logger.exception(f"Tool: {tool_name} FAILED", {
                "user_id": self._user_id,
                "duration_ms": _ms,
                "error": str(exc),
            })
            raise

    # Reminders
    async def _set_reminder(self, inp: dict[str, Any]) -> ToolResult:
        message = str(inp.get("message", "")).strip()
        delay_minutes = inp.get("delay_minutes")
        priority = str(inp.get("priority", "normal"))

        if not message:
            raise ValueError("message is required")
        if not isinstance(delay_minutes, (int, float)) or delay_minutes <= 0:
            raise ValueError("delay_minutes must be a positive number")

        trigger_at = (
            datetime.now(timezone.utc) + timedelta(minutes=float(delay_minutes))
        ).isoformat()

        reminder_id = str(uuid4())
        now_iso = datetime.now(timezone.utc).isoformat()

        data = {
            "id": reminder_id,
            "message": message,
            "trigger_at": trigger_at,
            "status": "pending",
            "priority": priority,
            "created_via": "voice",
            "snooze_count": 0,
            "created_at": now_iso,
        }
        ref = self._reminders_ref().document(reminder_id)
        await _run(lambda: ref.set(data))

        return {
            "reminder_id": reminder_id,
            "message": message,
            "trigger_at": trigger_at,
            "status": "pending",
            "priority": priority,
        }

    async def _list_reminders(self, inp: dict[str, Any]) -> ToolResult:
        status_filter = str(inp.get("status_filter", "pending"))

        def _fetch() -> list[dict]:
            q = self._reminders_ref().order_by("trigger_at")
            if status_filter != "all":
                q = q.where(filter=FieldFilter("status", "==", status_filter))
            return [{"reminder_id": d.id, **d.to_dict()} for d in q.stream()]

        reminders = await _run(_fetch)
        return {"reminders": reminders}

    async def _cancel_reminder(self, inp: dict[str, Any]) -> ToolResult:
        reminder_id = str(inp.get("reminder_id", "")).strip()
        if not reminder_id:
            raise ValueError("reminder_id is required")

        now_iso = datetime.now(timezone.utc).isoformat()
        ref = self._reminders_ref().document(reminder_id)
        await _run(lambda: ref.update({
            "status": "dismissed",
            "dismissed_at": now_iso,
        }))
        return {"reminder_id": reminder_id, "status": "dismissed"}

    # Calendar
    async def _create_calendar_event(self, inp: dict[str, Any]) -> ToolResult:
        title = str(inp.get("title", "")).strip()
        start_time = str(inp.get("start_time", "")).strip()
        if not title or not start_time:
            raise ValueError("title and start_time are required")

        end_time = inp.get("end_time")
        if not end_time:
            start_dt = datetime.fromisoformat(start_time)
            end_time = (start_dt + timedelta(minutes=30)).isoformat()

        body: dict[str, Any] = {
            "summary": title,
            "start": {"dateTime": start_time},
            "end": {"dateTime": end_time},
        }
        if inp.get("description"):
            body["description"] = inp["description"]
        if inp.get("location"):
            body["location"] = inp["location"]

        def _create() -> ToolResult:
            connector = GoogleCalendarConnector(self._user_id)
            status = connector.get_status()
            if not status.get("enabled"):
                return {"configured": False, "message": "Google Calendar is not configured."}
            cal = connector.calendar_client()
            event = cal.events().insert(calendarId="primary", body=body).execute()
            connector.cache_api_events([event])
            return {
                "configured": True,
                "event_id": event.get("id"),
                "html_link": event.get("htmlLink"),
                "status": event.get("status"),
            }

        return await _run(_create)

    async def _get_upcoming_events(self, inp: dict[str, Any]) -> ToolResult:
        def _fetch() -> ToolResult:
            connector = GoogleCalendarConnector(self._user_id)
            return connector.query_events(
                range_name=str(inp.get("range", "")).strip() or None,
                start_time=str(inp.get("start_time", "")).strip() or None,
                end_time=str(inp.get("end_time", "")).strip() or None,
                limit=int(inp.get("limit", 10) or 10),
                hours_ahead=int(inp.get("hours_ahead", 24) or 24),
            )

        return await _run(_fetch)

    # Memory
    async def _store_memory(self, inp: dict[str, Any]) -> ToolResult:
        key = str(inp.get("key", "")).strip()
        value = str(inp.get("value", "")).strip()
        category = str(inp.get("category", "")).strip()

        if not key or not value or not category:
            raise ValueError("key, value, and category are required")

        now_iso = datetime.now(timezone.utc).isoformat()

        def _upsert() -> str:
            existing = list(
                self._memories_ref().where(filter=FieldFilter("key", "==", key)).limit(1).stream()
            )
            if existing:
                memory_id = existing[0].id
                self._memories_ref().document(memory_id).set(
                    {"key": key, "value": value, "category": category, "updated_at": now_iso},
                    merge=True,
                )
            else:
                memory_id = str(uuid4())
                self._memories_ref().document(memory_id).set({
                    "key": key,
                    "value": value,
                    "category": category,
                    "source": "voice",
                    "created_at": now_iso,
                    "updated_at": now_iso,
                })
            return memory_id

        memory_id = await _run(_upsert)
        return {"memory_id": memory_id, "key": key, "value": value, "category": category}

    async def _query_memory(self, inp: dict[str, Any]) -> ToolResult:
        query_str = str(inp.get("query", "")).strip().lower()
        category_filter = str(inp.get("category_filter", "all"))

        if not query_str:
            raise ValueError("query is required")

        def _search() -> list[dict]:
            q = self._memories_ref()
            if category_filter != "all":
                q = q.where(filter=FieldFilter("category", "==", category_filter))
            matches: list[dict] = []
            for doc in q.stream():
                data = doc.to_dict() or {}
                haystack = f"{data.get('key', '')} {data.get('value', '')}".lower()
                if query_str in haystack:
                    matches.append({"memory_id": doc.id, **data})
                if len(matches) >= 10:
                    break
            return matches

        matches = await _run(_search)
        return {"matches": matches}

    # Nutrition
    async def _analyze_nutrition(self, inp: dict[str, Any]) -> ToolResult:
        ocr_text = str(inp.get("ocr_text", "")).strip()
        if not ocr_text:
            raise ValueError("ocr_text is required")

        quantity = float(inp.get("quantity", 1) or 1)

        def _extract(pattern: str) -> float:
            m = re.search(pattern, ocr_text, re.IGNORECASE)
            return float(m.group(1)) if m else 0.0

        calories = _extract(r"calories[^\d]*(\d+(?:\.\d+)?)")
        protein = _extract(r"protein[^\d]*(\d+(?:\.\d+)?)")
        sugar = _extract(r"sugar[^\d]*(\d+(?:\.\d+)?)")
        sodium = _extract(r"sodium[^\d]*(\d+(?:\.\d+)?)")

        concerns: list[str] = []
        if sugar * quantity >= 20:
            concerns.append("high sugar")
        if sodium * quantity >= 600:
            concerns.append("high sodium")
        if protein * quantity <= 5:
            concerns.append("low protein")

        recommendation = "moderate" if ("high sugar" in concerns or "high sodium" in concerns) else "eat"

        log_id = str(uuid4())
        now_iso = datetime.now(timezone.utc).isoformat()
        log_data = {
            "ocr_text": ocr_text,
            "occasion": inp.get("occasion"),
            "quantity": quantity,
            "is_cheat_meal": bool(inp.get("is_cheat_meal", False)),
            "analysis": {"calories": calories, "protein": protein, "sugar": sugar, "sodium": sodium},
            "concerns": concerns,
            "recommendation": recommendation,
            "timestamp": now_iso,
        }
        ref = self._nutrition_logs_ref().document(log_id)
        await _run(lambda: ref.set(log_data))

        return {
            "nutrition_log_id": log_id,
            "calories": calories,
            "protein_grams": protein,
            "sugar_grams": sugar,
            "sodium_mg": sodium,
            "quantity": quantity,
            "concerns": concerns,
            "recommendation": recommendation,
        }

    # Clarification (chat-only — returns sentinel dict, not a Firestore call)
    async def _ask_clarification(self, inp: dict[str, Any]) -> ToolResult:
        return {
            "__clarification__": True,
            "clarification_id": str(uuid4()),
            "question": str(inp.get("question", "")).strip(),
            "options": [str(o) for o in inp.get("options", [])],
            "multi_select": bool(inp.get("multi_select", False)),
        }

    # User context
    async def _get_user_context(self, inp: dict[str, Any]) -> ToolResult:
        include_memories = bool(inp.get("include_memories", True))
        include_reminders = bool(inp.get("include_reminders", True))
        include_events = bool(inp.get("include_events", True))

        context: dict[str, Any] = {"user_id": self._user_id}

        if include_memories:
            context["memories"] = await _run(
                lambda: [{"memory_id": d.id, **d.to_dict()} for d in self._memories_ref().stream()]
            )

        if include_reminders:
            context["reminders"] = await _run(
                lambda: [
                    {"reminder_id": d.id, **d.to_dict()}
                    for d in self._reminders_ref().where(filter=FieldFilter("status", "==", "pending")).stream()
                ]
            )

        if include_events:
            result = await self._get_upcoming_events({"hours_ahead": 24})
            context["upcoming_events"] = result.get("events", [])

        return context


# Standalone Firestore helpers (used by scheduler)
def fetch_due_reminders() -> list[dict[str, Any]]:
    """Query all users' pending reminders that are due now.

    Intentionally synchronous — called via asyncio.to_thread from the scheduler.
    """                 
    db = admin_firestore()
    now_iso = datetime.now(timezone.utc).isoformat()

    docs = (
        db.collection_group("reminders")
        .where(filter=FieldFilter("status", "==", "pending"))
        .where(filter=FieldFilter("trigger_at", "<=", now_iso))
        .stream()
    )

    results = []
    for doc in docs:
        parent = doc.reference.parent.parent
        if parent is None:
            logger.error("Could not resolve userId for reminder", {"doc_id": doc.id})
            continue
        results.append({"userId": parent.id, "reminderId": doc.id, "data": doc.to_dict()})
    return results


def mark_reminder_fired(user_id: str, reminder_id: str) -> None:
    """Intentionally synchronous — called via asyncio.to_thread from the scheduler."""
    db = admin_firestore()
    now_iso = datetime.now(timezone.utc).isoformat()
    db.collection("users").document(user_id).collection("reminders").document(reminder_id).update({
        "status": "fired",
        "fired_at": now_iso,
    })


def list_user_fcm_tokens(user_id: str) -> list[str]:
    """Return all FCM token strings for a user.

    Reads from the ``users/{uid}/fcm_tokens`` subcollection managed by
    :mod:`fcm_token_registry`.  Kept for backward compatibility with any
    callers that haven't been migrated to ``send_notification`` yet.
    """
    from .fcm_token_registry import get_user_tokens
    return [t["token"] for t in get_user_tokens(user_id)]


def log_tool_failure(tool_name: str, error: Exception) -> None:
    logger.error("Tool execution failed", {"tool": tool_name, "error": str(error)})
